### DISCUSSION TITLE
Any scripts/plugins for better JSDoc comment formatting?

### DISCUSSION THREAD

**Discussion by trusktr** [neovim/neovim#13725] [Q&A] [answered]
At the moment my JSDoc formatting is horrendous.

For example, looks what happens when I type a comment, select it, and pre `gq`:

https://user-images.githubusercontent.com/297678/104081094-5924af80-51e1-11eb-8276-fe8f8fd711ff.mp4

Is there a way to improve comment formatting (I mean, while typing and editing and using things like `gq`)?

For now, I format with Prettier on each save, so that at least indents everything correctly without hassle, but `gq` can be really bad.

**Comment by mjlbach** [comment, marked as answer]
This is already built-in to vim with formatprg

```
au FileType javascript setlocal formatprg=prettier
au FileType javascript.jsx setlocal formatprg=prettier
au FileType typescript setlocal formatprg=prettier\ — parser\ typescript
```

**Reply by trusktr** [reply]
I made sure `prettier` is in my `PATH`, added those autocmds, and then I tried `gq` on a comment line:


https://user-images.githubusercontent.com/297678/104081873-8162dd00-51e6-11eb-82f7-578a71c4f05f.mp4



I don't think Prettier formats inside of comments anyway. It will only indent the lines after I run the file through prettier, like so:


https://user-images.githubusercontent.com/297678/104081888-9b9cbb00-51e6-11eb-924b-e2811a3ebccf.mp4

But running `gq` on those two lines brings it back:


https://user-images.githubusercontent.com/297678/104081898-af482180-51e6-11eb-89e2-8d7c35d800c4.mp4

**Reply by trusktr** [reply]
Thanks for mentioning `formatprg` I was aware of that (or I forgot, it's been a while since I configured vim, and have been using vim emulators like the excellent vim-mode-plus in Atom which is _so good_ it is effectively a complete alternative to Vim or Neovim but with all scripting in JavaScript, and the default formatters in Atom/VSCode/etc are very good by default).

**Reply by trusktr** [reply]
I wonder if maybe there's a formatter specifically for C-like comments. But how would it mesh with a formatter for the rest of the file? It's a bit tricky it seems.

**Reply by trusktr** [reply]
Sorry, I was wrong! I am running Neovim inside of VS Code, but apparently `gq` tells Neovim to tell VS Code to format the wrapping instead, and in that case it doesn't work, but it is fine in plain Neovim outside of VS Code.

Also I had to restart my terminal, not just neovim (even though `prettier` was in `PATH`) for it to show up in Neovim.

After that, and making sure `formatoptions` is set, it is working great.

**Comment by David-Else** [comment]
@mjlbach Hello. I have been trying all day to make Prettier format js/ts. What you pasted above seems to have typos, if you meant the following (they all need `----parser` to work):
```
  au FileType javascript setlocal formatprg=prettier\ --parser\ typescript
  au FileType javascript.jsx setlocal formatprg=prettier\ --parser\ typescript
  au FileType typescript setlocal formatprg=prettier\ --parser\ typescript
```
The `.js` files seem to work, but `.ts` acts as if there is no `formatprg` setting. It just bunches all the code to the top, as per default. Do you have an idea what could be going wrong? I am using a recent `nvim 0.5` nightly.

**Reply by mjlbach** [reply]
I don't really do TS/JS, I use formatprg generally and just copied an example off medium. Neovim shouldn't be doing anything different from vim in this regard, so I doubt the version that you are using will matter.

The following example if from the [prettier docs](https://prettier.io/docs/en/vim.html), so I don't think you *need* parser (unless it's a TS specific thing).

```vim
autocmd FileType javascript setlocal formatprg=prettier\ --single-quote\ --trailing-comma\ es5
" Use formatprg when available
let g:neoformat_try_formatprg = 1
```

I won't be much help, just wanted to point out the built-in vim functionality since OP was asking about plugins.
